addons/mail/static/src/core/common/message_model.js

```diff
@@ -40,6 +44,23 @@ export class Message extends Record {
     attachments = Record.many("Attachment", { inverse: "message" });
     author = Record.one("Persona");
     body = Record.attr("", { html: true });
+    bodyHtml = Record.attr("", {
+        compute() {
+            if (!this.isBodyEmpty) {
+                const htmlDoc = parser.parseFromString(this.body, "text/html");
+                const markdownElements = htmlDoc.querySelectorAll("odoo-markdown");
+                if (!this.store.markdownLoaded && markdownElements.length > 0) {
+                    this.store.loadMarkdownBundle();
+                    return this.body;
```

Comment:
To be improved, at least empty is be better than flicker of content.
To consider: not rendering message at all while this is not done.

Reply:
> To consider: not rendering message at all while this is not done.

I initially wanted to do this, but @alexkuhn doesn't want to delay the rendering.
I still think markdown should be loaded earlier, when discuss is "started" when a channel is open, or something like that.

Reply:
@phenix-factory I'm not against delaying the rendering, but it depends on how you implement it. If this is a `onWillStart(() => loadMarkdownBundle())` then this is not good, because this blocks parented rendering too.
If this is a `t-if` in `<Thread>` then why not.

We could fine-tune the moment we load markdown, but this is not a motivation to load it so early that we don't have to deal with async loading while we want to show a message. "When discuss is started" is at page load for setup of services and models and this is not ok because we want lazy-load. At channel opening is not much better, because a message is not necessarily shown there (messaging menu item, chat bubble preview, notif toaster, etc.). At least being in the use of `message.body` that contains some markdown ensures exact moment we need it and works in all current (and future) showing of message body in code.

Reply:
And not rendering the message is fine, but functionally speaking we don't really care. Sure this is theoretically less work for OWL, but this doesn't matter in terms of any perceivable performance gains.

Reply:
I'm experimenting with web workers, maybe this will become a nonissue.

Reply:
This is quite complex, I was going to post the following message (I still keep it below, because it is good food for thought) but then I realized it's quite confusing to display a channel (if it's empty) and then no longer display it while markdown is loading when a message is posted or if it receives a message.

And it's even more complicated with inbox.

Isn't it core enough that we shouldn't do lazy loading?
If we have discuss we almost certainly have channels (general), if we have channels we more than likely have messages, so either way we need the lib all the time.

Unless we only load it when a message is really displayed, which is the only way to lazy load it in a way that makes sense, but then we have all the issues I mentioned to resolve (flicker, combination with other features, etc).

---

I think it should be a condition _before_ rendering thread component, so in parents: discuss, public page, chat window. Messaging menu is making exception as it only displays the "preview", so it can be handled differently.

If the there is at least one markdown message inside the thread, it should wait on markdown being loaded before displaying the component.

We absolutely can't render empty thread/messages or thread with non-parsed markdown because it will mess up with load older, scroll restore, and stuff like that, due to them being heavily based on timings and rendering. We could refactor them to take delayed loading but it's probably more work?

The "trigger" for when loading the lib could be based on model state like it is now, it doesn't need to wait on the message being actually displayed. Knowledge in store of a message with markdown is more than enough to assume this message will soon enough be rendered.

Reply:
If most of the size is for code formatting, we should immediately load the parsing lib all the time, and only lazy-load code syntax when a specific language is used.